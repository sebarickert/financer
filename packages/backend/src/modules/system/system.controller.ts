import { Controller, DefaultValuePipe, Get, Query } from '@nestjs/common';
import { Role } from '@prisma/client';

import { Auth } from '../auth/decorators/auht.decorator';

import { SystemService } from './system.service';

@Controller('api/system')
@Auth(Role.ADMIN)
export class SystemController {
  constructor(private readonly systemService: SystemService) {}

  private static getMonthAgo(): Date {
    const date = new Date();
    date.setMonth(date.getMonth() - 1);
    return date;
  }

  @Get('logs')
  async getLogs(
    @Query('start', new DefaultValuePipe(SystemController.getMonthAgo()))
    start: Date,
    @Query('end', new DefaultValuePipe(new Date())) end: Date,
  ) {
    return this.systemService.getLogs(start, end);
  }
}
