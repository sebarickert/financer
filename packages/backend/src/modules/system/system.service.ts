import { Injectable } from '@nestjs/common';
import { SystemLog } from '@prisma/client';

import { SystemLogRepo } from '../../database/repos/system-log.repo';

import { CreateSystemLogDto } from './dto/create-system-log.dto';

@Injectable()
export class SystemService {
  constructor(private readonly systemLogRepo: SystemLogRepo) {}

  async getLogs(startDate: Date, endDate: Date): Promise<SystemLog[]> {
    const start =
      startDate.toString() !== 'Invalid Date'
        ? startDate
        : new Date(Date.now() - 30 * 24 * 60 * 60 * 1000); // 30 days

    const end = endDate.toString() !== 'Invalid Date' ? endDate : new Date();

    return this.systemLogRepo.findMany({
      where: {
        createdAt: { gte: start, lte: end },
      },
      orderBy: { createdAt: 'desc' },
    });
  }

  async createSystemLogEntry(logEntry: CreateSystemLogDto) {
    return this.systemLogRepo.create(logEntry);
  }
}
