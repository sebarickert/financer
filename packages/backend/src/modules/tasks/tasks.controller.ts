import { Controller, Get } from '@nestjs/common';

import { TasksService } from './tasks.service';

@Controller('api/tasks')
export class TasksController {
  constructor(private readonly tasksService: TasksService) {}

  @Get('generate-transactions')
  async generateTransactions() {
    return this.tasksService.generateTransactions();
  }
}
