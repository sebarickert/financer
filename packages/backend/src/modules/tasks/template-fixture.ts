export const templateFixture = {
  user: {
    roles: ['test-user'],
    _id: '61460d7354ea082ad0256749',
    name: 'Dummy Test User',
  },
  accounts: [
    {
      _id: '61460da354ea082ad025676b',
      name: 'Cash account',
      type: 'cash',
      balance: 11201.259999999995,
      owner: '61460d7354ea082ad0256749',
      isDeleted: false,
      __v: 0,
    },
  ],
  accountBalanceChanges: [],
  transactions: [],
  transactionCategories: [
    {
      _id: '623b58ada3deba9879422fbf',
      owner: '61460d7354ea082ad0256749',
      name: 'Category for all types',
      visibility: ['income', 'expense', 'transfer'],
      parent_category_id: null,
      deleted: false,
      __v: 0,
    },
  ],
  transactionCategoryMappings: [],
  userPreferences: [],
  transactionTemplates: [
    {
      _id: '638da83f4ce377d764461250',
      userId: '61460d7354ea082ad0256749',
      templateName: 'Test template',
      templateType: ['auto'],
      templateVisibility: 'income',
      fromAccount: null,
      toAccount: '61460da354ea082ad025676b',
      amount: 100,
      description: 'Test template',
      dayOfMonth: 15,
      dayOfMonthToCreate: 30,
      categories: [],
      __v: 0,
    },
    {
      _id: '638da86a4ce377d764461255',
      userId: '61460d7354ea082ad0256749',
      templateName: 'Test template 2',
      templateType: ['auto'],
      templateVisibility: 'income',
      fromAccount: null,
      toAccount: '61460da354ea082ad025676b',
      amount: 200,
      description: 'Test template 2',
      dayOfMonth: 31,
      dayOfMonthToCreate: 2,
      categories: [],
      __v: 0,
    },
  ],
};
