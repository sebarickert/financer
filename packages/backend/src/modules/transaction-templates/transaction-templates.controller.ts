import {
  Controller,
  Get,
  Post,
  Body,
  Patch,
  Param,
  Delete,
} from '@nestjs/common';
import { TransactionTemplateType } from '@prisma/client';
import {
  ApiBody,
  ApiOkResponse,
  ApiParam,
  ApiTags,
} from '@silte/nestjs-swagger';

import { ValidateEntityId } from '../../utils/validate-entity-id.pipe';
import { LoggedIn } from '../auth/decorators/loggedIn.decorators';
import { UserId } from '../users/users.decorators';

import { CreateTransactionTemplateDto } from './dto/create-transaction-template.dto';
import { TransactionTemplateDto } from './dto/transaction-template.dto';
import { UpdateTransactionTemplateDto } from './dto/update-transaction-template.dto';
import { TransactionTemplatesService as TransactionTemplatesService } from './transaction-templates.service';

@Controller('api/transaction-templates')
@ApiTags('Transaction templates')
@LoggedIn()
export class TransactionTemplatesController {
  constructor(
    private readonly transactionTemplatesService: TransactionTemplatesService,
  ) {}

  @Post()
  @ApiBody({ type: CreateTransactionTemplateDto })
  @ApiOkResponse({ schema: { properties: { payload: { type: 'string' } } } })
  create(
    @UserId() userId: string,
    @Body() createTransactionTemplateDto: CreateTransactionTemplateDto,
  ) {
    return this.transactionTemplatesService.create(
      createTransactionTemplateDto,
      userId,
    );
  }

  @Get()
  @ApiOkResponse({
    type: [TransactionTemplateDto],
    description: 'Return all transaction templates',
  })
  findAllByUser(@UserId() userId: string) {
    return this.transactionTemplatesService.findAllByUser(userId);
  }

  @Get('/manual')
  @ApiOkResponse({
    type: [TransactionTemplateDto],
    description: 'Return all transaction templates of type manual',
  })
  findAllManualTypeByUser(@UserId() userId: string) {
    return this.transactionTemplatesService.findAllByUserAndType(
      userId,
      TransactionTemplateType.MANUAL,
    );
  }

  @Get(':id')
  @ApiOkResponse({
    type: TransactionTemplateDto,
    description: 'Return transaction template by id',
  })
  @ApiParam({
    name: 'id',
    type: String,
  })
  findOne(@UserId() userId: string, @Param('id', ValidateEntityId) id: string) {
    return this.transactionTemplatesService.findOne(id, userId);
  }

  @Patch(':id')
  @ApiBody({ type: UpdateTransactionTemplateDto })
  @ApiOkResponse({ type: TransactionTemplateDto })
  @ApiParam({
    name: 'id',
    type: String,
  })
  update(
    @UserId() userId: string,
    @Param('id', ValidateEntityId) id: string,
    @Body() updateTransactionTemplateDto: UpdateTransactionTemplateDto,
  ) {
    return this.transactionTemplatesService.update(
      id,
      updateTransactionTemplateDto,
      userId,
    );
  }

  @Delete(':id')
  @ApiParam({
    name: 'id',
    type: String,
  })
  remove(@UserId() userId: string, @Param('id', ValidateEntityId) id: string) {
    return this.transactionTemplatesService.remove(id, userId);
  }
}
