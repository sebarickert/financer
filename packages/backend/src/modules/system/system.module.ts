import { Module } from '@nestjs/common';

import { DatabaseModule } from '../../database/database.module';

import { SystemController } from './system.controller';
import { SystemService } from './system.service';

@Module({
  imports: [DatabaseModule],
  controllers: [SystemController],
  providers: [SystemService],
  exports: [SystemService],
})
export class SystemModule {}
