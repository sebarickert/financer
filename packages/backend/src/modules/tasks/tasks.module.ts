import { Module } from '@nestjs/common';

import { TransactionTemplateModule } from '../transaction-templates/transaction-templates.module';
import { TransactionsModule } from '../transactions/transactions.module';

import { TasksController } from './tasks.controller';
import { TasksService } from './tasks.service';

@Module({
  imports: [TransactionTemplateModule, TransactionsModule],
  controllers: [TasksController],
  providers: [TasksService],
})
export class TasksModule {}
