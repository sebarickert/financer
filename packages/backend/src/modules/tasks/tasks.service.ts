import { Injectable } from '@nestjs/common';
import { SystemLogLevel, TransactionTemplateType } from '@prisma/client';

import { parseObjectId } from '../../types/objectId';
import { getLastDayOfMonth } from '../../utils/date-utils';
import { SystemService } from '../system/system.service';
import { TransactionTemplatesService } from '../transaction-templates/transaction-templates.service';
import { TransactionsService } from '../transactions/transactions.service';

@Injectable()
export class TasksService {
  constructor(
    private readonly templateService: TransactionTemplatesService,
    private readonly transactionsService: TransactionsService,
    private readonly systemService: SystemService,
  ) {}

  async generateTransactions() {
    const now = new Date();
    const dayOfMonth = now.getDate();

    const templates =
      await this.templateService.findAutomatedTemplatesWithCreationDateBefore(
        dayOfMonth,
        dayOfMonth < getLastDayOfMonth() ? 'eq' : 'gte',
      );

    const templateLogEntries =
      await this.templateService.findTemplateLogEntriesByTemplateIdsAndType(
        templates.map((template) => template.id),
        TransactionTemplateType.AUTO,
      );

    const result = await Promise.all(
      templates.map(async (template) => {
        if (!template.amount || !template.description) return 'missingData';

        const hasAddedTransactionThisMonth = templateLogEntries.some(
          (entry) => {
            if (entry.templateId !== template.id) return false;

            const entryDate = new Date(entry.executed);
            return (
              entryDate.getMonth() === now.getMonth() &&
              entryDate.getFullYear() === now.getFullYear()
            );
          },
        );

        if (hasAddedTransactionThisMonth) return 'skipped';

        const transactionData =
          this.templateService.getTransactionFromTemplate(template);
        const userId = template.userId;

        const transaction = await this.transactionsService.create(
          parseObjectId(userId),
          transactionData,
        );

        await this.templateService.createTemplateLogEntry({
          templateId: template.id,
          transactionId: transaction._id.toString(),
          executed: now,
          userId,
          eventType: TransactionTemplateType.AUTO,
        });

        return 'added';
      }),
    );

    const output = result.reduce(
      (acc, curr) => {
        if (curr === 'skipped') {
          acc.skipped += 1;
        } else if (curr === 'missingData') {
          acc.missingData += 1;
        } else {
          acc.added += 1;
        }

        return acc;
      },
      { skipped: 0, added: 0, missingData: 0 },
    );

    await this.systemService.createSystemLogEntry({
      module: 'tasks',
      service: 'generateTransactions',
      message: `Added ${output.added} transactions, skipped ${output.skipped} and found ${output.missingData} templates with missing data`,
      level: SystemLogLevel.INFO,
    });

    return output;
  }
}
