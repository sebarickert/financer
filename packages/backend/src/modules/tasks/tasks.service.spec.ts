import { TransactionType } from '@local/types';
import { Test, TestingModule } from '@nestjs/testing';

import { rootMongooseTestModule } from '../../../test/rootMongooseTest.module';
import { getMongoConnection } from '../../config/memoryDatabaseServer';
import { parseObjectId } from '../../types/objectId';
import { SystemModule } from '../system/system.module';
import { TransactionTemplateModule } from '../transaction-templates/transaction-templates.module';
import { TransactionsModule } from '../transactions/transactions.module';
import { TransactionsService } from '../transactions/transactions.service';
import { UserDataModule } from '../user-data/user-data.module';
import { UserDataService } from '../user-data/user-data.service';

import { TasksService } from './tasks.service';
import { templateFixture } from './template-fixture';

const dummyUserId = parseObjectId('61460d7354ea082ad0256749');

describe('TasksService', () => {
  let service: TasksService;
  let userDataservice: UserDataService;
  let transactionsService: TransactionsService;

  let RealDate: DateConstructor;
  let dateMock: Date;

  const mockDate = () => {
    RealDate = Date;
    dateMock = new Date(2022, 8, 2, 12);
    jest.spyOn(global, 'Date').mockImplementation((...args) => {
      if (args.length) {
        return new RealDate(...args);
      }
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      return new RealDate(dateMock) as any;
    });

    ['prototype', 'now'].forEach(
      (methodName) => (Date[methodName] = RealDate[methodName]),
    );
    Date.now = () => dateMock.getTime();
  };

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [
        rootMongooseTestModule(),
        UserDataModule,
        TransactionTemplateModule,
        TransactionsModule,
        SystemModule,
      ],
      providers: [TasksService],
    }).compile();

    service = module.get<TasksService>(TasksService);
    userDataservice = module.get<UserDataService>(UserDataService);
    transactionsService = module.get<TransactionsService>(TransactionsService);

    mockDate();
  });

  afterEach(async () => {
    const connection = await getMongoConnection();
    const collections = await connection.db.collections();
    await Promise.all(
      collections.map((collection) => collection.deleteMany({})),
    );
    await connection.close();

    global.Date = RealDate;
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should create transaction from template', async () => {
    await userDataservice.overrideUserData(dummyUserId, {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      ...(templateFixture as any),
    });

    const transactionsBefore = await transactionsService.findAllByUser(
      dummyUserId,
      TransactionType.ANY,
    );

    expect(transactionsBefore.data).toHaveLength(0);

    await service.generateTransactions();

    const transactionsAfter = await transactionsService.findAllByUser(
      dummyUserId,
      TransactionType.ANY,
    );

    expect(transactionsAfter.data).toHaveLength(1);
  });

  it('should not create duplicated transaction from template with multiple executions', async () => {
    await userDataservice.overrideUserData(dummyUserId, {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      ...(templateFixture as any),
    });

    const transactionsBefore = await transactionsService.findAllByUser(
      dummyUserId,
      TransactionType.ANY,
    );

    expect(transactionsBefore.data).toHaveLength(0);

    await service.generateTransactions();
    await service.generateTransactions();

    const transactionsAfter = await transactionsService.findAllByUser(
      dummyUserId,
      TransactionType.ANY,
    );

    expect(transactionsAfter.data).toHaveLength(1);
  });

  it('should use last day of month id dayOfMonth is higher than amount of days', async () => {
    dateMock = new Date(2022, 1, 2, 12);

    await userDataservice.overrideUserData(dummyUserId, {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      ...(templateFixture as any),
    });

    const transactionsBefore = await transactionsService.findAllByUser(
      dummyUserId,
      TransactionType.ANY,
    );

    expect(transactionsBefore.data).toHaveLength(0);

    await service.generateTransactions();

    const transactionsAfter = await transactionsService.findAllByUser(
      dummyUserId,
      TransactionType.ANY,
    );

    expect(transactionsAfter.data).toHaveLength(1);

    const expectedDate = new Date(2022, 1, 28, 12);
    const transactionDate = transactionsAfter.data[0].date;

    expect(transactionDate.getFullYear()).toBe(expectedDate.getFullYear());
    expect(transactionDate.getMonth()).toBe(expectedDate.getMonth());
    expect(transactionDate.getDate()).toBe(expectedDate.getDate());
  });

  it('should create transaction to next month if dayOfMonthToCreate is greater than dayOfMonth', async () => {
    dateMock = new Date(2022, 0, 30, 12);

    await userDataservice.overrideUserData(dummyUserId, {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      ...(templateFixture as any),
    });

    const transactionsBefore = await transactionsService.findAllByUser(
      dummyUserId,
      TransactionType.ANY,
    );

    expect(transactionsBefore.data).toHaveLength(0);

    await service.generateTransactions();

    const transactionsAfter = await transactionsService.findAllByUser(
      dummyUserId,
      TransactionType.ANY,
    );

    expect(transactionsAfter.data).toHaveLength(1);

    const expectedDate = new Date(2022, 1, 15, 12);
    const transactionDate = transactionsAfter.data[0].date;

    expect(transactionDate.getFullYear()).toBe(expectedDate.getFullYear());
    expect(transactionDate.getMonth()).toBe(expectedDate.getMonth());
    expect(transactionDate.getDate()).toBe(expectedDate.getDate());
  });

  it('should run templates on last day of month even with higher dayOfMonthToCreate values', async () => {
    dateMock = new Date(2022, 1, 28, 12);

    await userDataservice.overrideUserData(dummyUserId, {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      ...(templateFixture as any),
    });

    const transactionsBefore = await transactionsService.findAllByUser(
      dummyUserId,
      TransactionType.ANY,
    );

    expect(transactionsBefore.data).toHaveLength(0);

    await service.generateTransactions();

    const transactionsAfter = await transactionsService.findAllByUser(
      dummyUserId,
      TransactionType.ANY,
    );

    expect(transactionsAfter.data).toHaveLength(1);

    const expectedDate = new Date(2022, 2, 15, 12);
    const transactionDate = transactionsAfter.data[0].date;

    expect(transactionDate.getFullYear()).toBe(expectedDate.getFullYear());
    expect(transactionDate.getMonth()).toBe(expectedDate.getMonth());
    expect(transactionDate.getDate()).toBe(expectedDate.getDate());
  });
});
