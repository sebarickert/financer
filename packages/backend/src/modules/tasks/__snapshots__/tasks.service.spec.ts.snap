// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksService should create transaction from template 1`] = `
[
  {
    "amount": "200",
    "date": 2022-09-30T09:00:00.000Z,
    "description": "Test template 2",
    "fromAccount": null,
    "toAccount": "61460da354ea082ad025676b",
    "userId": "61460d7354ea082ad0256749",
  },
]
`;

exports[`TasksService should create transaction from template 2`] = `
[
  {
    "level": "INFO",
    "message": "Added 1 transactions, skipped 0 and found 0 templates with missing data",
    "module": "tasks",
    "service": "generateTransactions",
  },
]
`;

exports[`TasksService should create transaction to next month if dayOfMonthToCreate is greater than dayOfMonth 1`] = `
[
  {
    "amount": "200",
    "date": 2022-01-31T10:00:00.000Z,
    "description": "Test template 2",
    "fromAccount": null,
    "toAccount": "61460da354ea082ad025676b",
    "userId": "61460d7354ea082ad0256749",
  },
]
`;

exports[`TasksService should create transaction to next month if dayOfMonthToCreate is greater than dayOfMonth 2`] = `
[
  {
    "level": "INFO",
    "message": "Added 1 transactions, skipped 0 and found 0 templates with missing data",
    "module": "tasks",
    "service": "generateTransactions",
  },
]
`;

exports[`TasksService should not create duplicated transaction from template with multiple executions 1`] = `
[
  {
    "level": "INFO",
    "message": "Added 0 transactions, skipped 1 and found 0 templates with missing data",
    "module": "tasks",
    "service": "generateTransactions",
  },
]
`;

exports[`TasksService should run templates on last day of month even with higher dayOfMonthToCreate values 1`] = `
[
  {
    "amount": "200",
    "date": 2022-02-28T10:00:00.000Z,
    "description": "Test template 2",
    "fromAccount": null,
    "toAccount": "61460da354ea082ad025676b",
    "userId": "61460d7354ea082ad0256749",
  },
]
`;

exports[`TasksService should run templates on last day of month even with higher dayOfMonthToCreate values 2`] = `
[
  {
    "level": "INFO",
    "message": "Added 1 transactions, skipped 0 and found 0 templates with missing data",
    "module": "tasks",
    "service": "generateTransactions",
  },
]
`;

exports[`TasksService should use last day of month id dayOfMonth is higher than amount of days 1`] = `
[
  {
    "amount": "200",
    "date": 2022-02-28T10:00:00.000Z,
    "description": "Test template 2",
    "fromAccount": null,
    "toAccount": "61460da354ea082ad025676b",
    "userId": "61460d7354ea082ad0256749",
  },
]
`;

exports[`TasksService should use last day of month id dayOfMonth is higher than amount of days 2`] = `
[
  {
    "level": "INFO",
    "message": "Added 1 transactions, skipped 0 and found 0 templates with missing data",
    "module": "tasks",
    "service": "generateTransactions",
  },
]
`;
